File: gofer_models/src/task_run.rs
use super::{epoch, task::Task, Variable};
use gofer_proto::task_run::{TaskRunState, TaskRunStatus};
use gofer_proto::task_run_status_reason;
use serde::{Deserialize, Serialize};
use strum::{Display, EnumString};

/// Since task runs are basically an abstraction over containers, this tells us
/// which state of progress the container is currently in.
#[derive(Debug, Display, EnumString, Serialize, Deserialize, PartialEq, Eq, Clone)]
pub enum State {
    /// Cannot determine state of task run, should never be in this state.
    Unknown,
    /// Task run is going through pre-scheduling verification and prep.
    Processing,
    /// Task run is waiting on parents task runs to finish.
    Waiting,
    /// Task run is currently running has reported by the scheduler.
    Running,
    /// Task run has completed.
    Complete,
}

impl From<State> for TaskRunState {
    fn from(r: State) -> Self {
        match r {
            State::Unknown => TaskRunState::UnknownState,
            State::Processing => TaskRunState::Processing,
            State::Waiting => TaskRunState::Waiting,
            State::Running => TaskRunState::Running,
            State::Complete => TaskRunState::Complete,
        }
    }
}

impl From<TaskRunState> for State {
    fn from(r: TaskRunState) -> Self {
        match r {
            TaskRunState::UnknownState => State::Unknown,
            TaskRunState::Processing => State::Processing,
            TaskRunState::Waiting => State::Waiting,
            TaskRunState::Running => State::Running,
            TaskRunState::Complete => State::Complete,
        }
    }
}

/// Since task runs are basically an abstraction over containers, this tells us
/// which status the container is in upon completion.
#[derive(Debug, Display, EnumString, Serialize, Deserialize, PartialEq, Eq, Clone)]
pub enum Status {
    /// Status is unknown; if task run is complete and this is that status, something is wrong.
    Unknown,
    /// The task run has completed successfully.
    Successful,
    /// The task run has failed either with an abnormal error code or during processing.
    Failed,
    /// The task run was cancelled during it's execution. Cancelled is explicitly a user
    /// invoked action. The only way a task gets cancelled is from an external request for
    /// it to be.
    Cancelled,
    /// The task run was skipped; This can happen be due to a task failing to meet it's dependencies
    /// (for instance it's parent was in an incorrect state).
    Skipped,
}

impl Default for Status {
    fn default() -> Self {
        Self::Unknown
    }
}

impl From<Status> for TaskRunStatus {
    fn from(r: Status) -> Self {
        match r {
            Status::Unknown => TaskRunStatus::UnknownStatus,
            Status::Successful => TaskRunStatus::Successful,
            Status::Failed => TaskRunStatus::Failed,
            Status::Cancelled => TaskRunStatus::Cancelled,
            Status::Skipped => TaskRunStatus::Skipped,
        }
    }
}

impl From<TaskRunStatus> for Status {
    fn from(r: TaskRunStatus) -> Self {
        match r {
            TaskRunStatus::UnknownStatus => Status::Unknown,
            TaskRunStatus::Successful => Status::Successful,
            TaskRunStatus::Failed => Status::Failed,
            TaskRunStatus::Cancelled => Status::Cancelled,
            TaskRunStatus::Skipped => Status::Skipped,
        }
    }
}

#[derive(Debug, Clone, Display, EnumString, Serialize, Deserialize, PartialEq, Eq)]
pub enum Reason {
    /// Gofer has no idea how the task run got into this state.
    Unknown,
    /// A non-zero exit code has been received.
    AbnormalExit,
    /// Encountered an error with the backend scheduler.
    SchedulerError,
    /// User error in task run parameters.
    FailedPrecondition,
    /// User invoked cancellation.
    Cancelled,
    /// Task run was lost due to internal error in tracking.
    Orphaned,
}

impl From<Reason> for task_run_status_reason::Reason {
    fn from(r: Reason) -> Self {
        match r {
            Reason::Unknown => task_run_status_reason::Reason::Unknown,
            Reason::AbnormalExit => task_run_status_reason::Reason::AbnormalExit,
            Reason::SchedulerError => task_run_status_reason::Reason::SchedulerError,
            Reason::FailedPrecondition => task_run_status_reason::Reason::FailedPrecondition,
            Reason::Cancelled => task_run_status_reason::Reason::Cancelled,
            Reason::Orphaned => task_run_status_reason::Reason::Orphaned,
        }
    }
}

impl From<task_run_status_reason::Reason> for Reason {
    fn from(r: task_run_status_reason::Reason) -> Self {
        match r {
            task_run_status_reason::Reason::Unknown => Reason::Unknown,
            task_run_status_reason::Reason::AbnormalExit => Reason::AbnormalExit,
            task_run_status_reason::Reason::SchedulerError => Reason::SchedulerError,
            task_run_status_reason::Reason::FailedPrecondition => Reason::FailedPrecondition,
            task_run_status_reason::Reason::Cancelled => Reason::Cancelled,
            task_run_status_reason::Reason::Orphaned => Reason::Orphaned,
        }
    }
}

/// A description of the current status of a task run.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct StatusReason {
    /// The kind of reason for the current status.
    pub reason: Reason,
    /// A short description of the reason.
    pub description: String,
}

impl From<StatusReason> for gofer_proto::TaskRunStatusReason {
    fn from(r: StatusReason) -> Self {
        Self {
            reason: task_run_status_reason::Reason::from(r.reason) as i32,
            description: r.description,
        }
    }
}

impl From<gofer_proto::TaskRunStatusReason> for StatusReason {
    fn from(r: gofer_proto::TaskRunStatusReason) -> Self {
        Self {
            reason: gofer_proto::task_run_status_reason::Reason::from_i32(r.reason)
                .unwrap()
                .into(),
            description: r.description,
        }
    }
}

/// A task run is a specific execution of a task/container.
/// It represents a 4th level unit in the hierarchy:
/// namespace -> pipeline -> run -> taskrun
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct TaskRun {
    /// Unique identifier for namespace.
    pub namespace: String,
    /// Unique identifier for pipeline.
    pub pipeline: String,
    /// Unique identifier for run.
    pub run: u64,
    /// Unique identifier for task run. Taken from the task identifier.
    pub id: String,
    /// The task information of the task associated with this particular task run.
    pub task: Task,
    /// Time the task run was created. Essentially whenever the run has started.
    pub created: u64,
    /// Time the task run was started on the scheduler.
    pub started: u64,
    /// Time the task run completed.
    pub ended: u64,
    /// The exit code of the task run.
    pub exit_code: Option<u8>,
    /// If the logs have past their predefined retention time.
    pub logs_expired: bool,
    /// If the logs have been removed due to user request or automatic action based on expiry time.
    pub logs_removed: bool,
    /// The current place of progress that task run is at.
    pub state: State,
    /// Upon completion of the task run, the status it has completed with.
    pub status: Status,
    /// Extra information about the current status.
    pub status_reason: Option<StatusReason>,
    /// Identifier used by the scheduler to identify this specific task run container.
    /// This is provided by the scheduler at the time of scheduling.
    pub scheduler_id: Option<String>,
    /// The environment variables injected during this particular task run.
    pub variables: Vec<Variable>,
}

impl TaskRun {
    pub fn new(namespace: &str, pipeline: &str, run: u64, task: Task) -> Self {
        Self {
            namespace: namespace.to_string(),
            pipeline: pipeline.to_string(),
            run,
            id: task.id.clone(),
            task,
            created: epoch(),
            started: 0,
            ended: 0,
            exit_code: None,
            status_reason: None,
            logs_expired: false,
            logs_removed: false,
            state: State::Processing,
            status: Status::Unknown,
            scheduler_id: None,
            variables: vec![],
        }
    }
}

impl From<TaskRun> for gofer_proto::TaskRun {
    fn from(r: TaskRun) -> Self {
        Self {
            namespace_id: r.namespace,
            pipeline_id: r.pipeline,
            run_id: r.run,
            id: r.id,
            task: Some(r.task.into()),
            created: r.created,
            started: r.started,
            ended: r.ended,
            exit_code: r.exit_code.unwrap_or_default() as u64,
            status_reason: r.status_reason.map(|r| r.into()),
            logs_expired: r.logs_expired,
            logs_removed: r.logs_removed,
            state: TaskRunState::from(r.state) as i32,
            status: TaskRunStatus::from(r.status) as i32,
            scheduler_id: r.scheduler_id.unwrap_or_default(),
            variables: r.variables.into_iter().map(|value| value.into()).collect(),
        }
    }
}

impl From<gofer_proto::TaskRun> for TaskRun {
    fn from(r: gofer_proto::TaskRun) -> Self {
        Self {
            namespace: r.namespace_id,
            pipeline: r.pipeline_id,
            run: r.run_id,
            id: r.id,
            task: r.task.unwrap().into(),
            created: r.created,
            started: r.started,
            ended: r.ended,
            exit_code: Some(r.exit_code as u8),
            status_reason: r.status_reason.map(|s| s.into()),
            logs_expired: r.logs_expired,
            logs_removed: r.logs_removed,
            state: gofer_proto::task_run::TaskRunState::from_i32(r.state)
                .unwrap()
                .into(),
            status: gofer_proto::task_run::TaskRunStatus::from_i32(r.status)
                .unwrap()
                .into(),
            scheduler_id: Some(r.scheduler_id),
            variables: r.variables.into_iter().map(|v| v.into()).collect(),
        }
    }
}

File: models/taskRun.go
package models

import (
	"time"

	proto "github.com/clintjedwards/gofer/proto/go"
)

type TaskRunState string

const (
	TaskRunStateUnknown    TaskRunState = "UNKNOWN"    // Unknown state, should never be in this state.
	TaskRunStateProcessing TaskRunState = "PROCESSING" // Pre-scheduling validation and prep.
	TaskRunStateWaiting    TaskRunState = "WAITING"    // Waiting to be scheduled.
	TaskRunStateRunning    TaskRunState = "RUNNING"    // Currently running as reported by scheduler.
	TaskRunStateComplete   TaskRunState = "COMPLETE"
)

type TaskRunStatus string

const (
	TaskRunStatusUnknown    TaskRunStatus = "UNKNOWN"
	TaskRunStatusFailed     TaskRunStatus = "FAILED"     // Has encountered an issue, either container issue or scheduling issue.
	TaskRunStatusSuccessful TaskRunStatus = "SUCCESSFUL" // Finished with a proper error code.
	TaskRunStatusCancelled  TaskRunStatus = "CANCELLED"  // Cancelled mid run due to user requested cancellation.
	TaskRunStatusSkipped    TaskRunStatus = "SKIPPED"    // Not run due to dependencies not being met.
)

type TaskRunStatusReasonKind string

const (
	TaskRunStatusReasonKindUnknown            TaskRunStatusReasonKind = "UNKNOWN"             // Unknown state, should never be in this state.
	TaskRunStatusReasonKindAbnormalExit       TaskRunStatusReasonKind = "ABNORMAL_EXIT"       // A non-zero exit code has been received.
	TaskRunStatusReasonKindSchedulerError     TaskRunStatusReasonKind = "SCHEDULER_ERROR"     // Encountered an error with the backend scheduler.
	TaskRunStatusReasonKindFailedPrecondition TaskRunStatusReasonKind = "FAILED_PRECONDITION" // User error in task run parameters.
	TaskRunStatusReasonKindCancelled          TaskRunStatusReasonKind = "CANCELLED"           // User invoked cancellation.
	TaskRunStatusReasonKindOrphaned           TaskRunStatusReasonKind = "ORPHANED"            // Task run was lost due to extreme internal error.
)

// A task run is a specific execution of a task/container.
// It represents a 4th level unit in the hierarchy:
//
//	namespace -> pipeline -> run -> taskrun
type TaskRun struct {
	Namespace   string `json:"namespace"`    // Unique identifier for namespace.
	Pipeline    string `json:"pipeline"`     // Unique pipeline ID of task run.
	Run         int64  `json:"run"`          // Unique run ID of task run; sequential.
	ID          string `json:"id"`           // Unique ID for task run; taken from the taskID.
	Created     int64  `json:"created"`      // Time of task run creation in epoch milliseconds.
	Started     int64  `json:"started"`      // Time of task run actual start in epoch milliseconds.
	Ended       int64  `json:"ended"`        // Time of task run completion in epoch milliseconds.
	ExitCode    *int64 `json:"exit_code"`    // The exit code of the task run.
	LogsExpired bool   `json:"logs_expired"` // If the logs have past their retention time.
	// If the logs have been removed. This can be due to user request or automatic action based on expiry time.
	LogsRemoved  bool                 `json:"logs_removed"`
	State        TaskRunState         `json:"state"`
	Status       TaskRunStatus        `json:"status"`
	StatusReason *TaskRunStatusReason `json:"status_reason"` // Extra information about the current status.
	Variables    []Variable           `json:"variables"`     // The environment variables injected during this particular task run.
	Task         Task                 `json:"task"`          // Task information.
}

type TaskRunStatusReason struct {
	Reason      TaskRunStatusReasonKind `json:"reason"`      // Specific type; useful for documentation.
	Description string                  `json:"description"` // Details about type.
}

func (t *TaskRunStatusReason) ToProto() *proto.TaskRunStatusReason {
	return &proto.TaskRunStatusReason{
		Reason:      proto.TaskRunStatusReason_Reason(proto.TaskRunStatusReason_Reason_value[string(t.Reason)]),
		Description: t.Description,
	}
}

func (t *TaskRunStatusReason) FromProto(proto *proto.TaskRunStatusReason) {
	t.Reason = TaskRunStatusReasonKind(proto.Reason.String())
	t.Description = proto.Description
}

func NewTaskRun(namespace, pipeline string, run int64, task Task) *TaskRun {
	return &TaskRun{
		Namespace:    namespace,
		Pipeline:     pipeline,
		Run:          run,
		ID:           task.GetID(),
		Created:      time.Now().UnixMilli(),
		Started:      0,
		Ended:        0,
		ExitCode:     nil,
		StatusReason: nil,
		LogsExpired:  false,
		LogsRemoved:  false,
		State:        TaskRunStateProcessing,
		Status:       TaskRunStatusUnknown,
		Variables:    []Variable{},
		Task:         task,
	}
}

func (r *TaskRun) ToProto() *proto.TaskRun {
	variables := []*proto.Variable{}
	for _, variable := range r.Variables {
		variables = append(variables, variable.ToProto())
	}

	var statusReason *proto.TaskRunStatusReason
	if r.StatusReason != nil {
		statusReason = r.StatusReason.ToProto()
	}

	var exitCode int64 = 155
	if r.ExitCode != nil {
		exitCode = *r.ExitCode
	}

	protoTaskRun := &proto.TaskRun{
		Namespace:    r.Namespace,
		Pipeline:     r.Pipeline,
		Run:          r.Run,
		Id:           r.ID,
		Created:      r.Created,
		Started:      r.Started,
		Ended:        r.Ended,
		ExitCode:     exitCode,
		StatusReason: statusReason,
		LogsExpired:  r.LogsExpired,
		LogsRemoved:  r.LogsRemoved,
		State:        proto.TaskRun_TaskRunState(proto.TaskRun_TaskRunState_value[string(r.State)]),
		Status:       proto.TaskRun_TaskRunStatus(proto.TaskRun_TaskRunStatus_value[string(r.Status)]),
		Variables:    variables,
	}

	switch task := r.Task.(type) {
	case *CommonTask:
		protoTaskRun.Task = &proto.TaskRun_CommonTask{
			CommonTask: task.ToProto(),
		}
	case *CustomTask:
		protoTaskRun.Task = &proto.TaskRun_CustomTask{
			CustomTask: task.ToProto(),
		}
	}

	return protoTaskRun
}

func (r *TaskRun) FromProto(pb *proto.TaskRun) {
	var statusReason *TaskRunStatusReason
	if pb.StatusReason != nil {
		sr := &TaskRunStatusReason{}
		sr.FromProto(pb.StatusReason)
		statusReason = sr
	}

	variables := []Variable{}
	for _, variable := range pb.Variables {
		vari := Variable{}
		vari.FromProto(variable)
		variables = append(variables, vari)
	}

	var task Task

	switch t := pb.Task.(type) {
	case *proto.TaskRun_CommonTask:
		commonTask := &CommonTask{}
		commonTask.FromProto(t.CommonTask)
		task = commonTask
	case *proto.TaskRun_CustomTask:
		customTask := &CustomTask{}
		customTask.FromProto(t.CustomTask)
		task = customTask
	}

	r.Namespace = pb.Namespace
	r.Pipeline = pb.Pipeline
	r.Run = pb.Run
	r.ID = pb.Id
	r.Created = pb.Created
	r.Started = pb.Started
	r.Ended = pb.Ended
	r.ExitCode = Ptr(pb.ExitCode)
	r.StatusReason = statusReason
	r.LogsExpired = pb.LogsExpired
	r.LogsRemoved = pb.LogsRemoved
	r.State = TaskRunState(pb.State.String())
	r.Status = TaskRunStatus(pb.Status.String())
	r.Variables = variables
	r.Task = task
}

func Ptr[T any](v T) *T {
	return &v
}
